pkg/controller/common.go

```diff
@@ -597,335 +610,806 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
 			if selectable {
-				podAffinities = podAffinities.Add(requirements...)
+				podAntiAffinities := labels.NewSelector().Add(requirements...)
+				pods := []v1.Pod{}
+
+				if len(podAntiAffinity.Namespaces) > 0 {
+					for _, namespace := range podAntiAffinity.Namespaces {
+						podList := &v1.PodList{}
+						if err = p.azr.getClient().List(ctx, podList, &client.ListOptions{LabelSelector: podAntiAffinities, Namespace: namespace}); err != nil {
+							klog.Errorf("failed to retrieve pod list: %v", err)
+							continue
+						}
+						pods = append(pods, podList.Items...)
+					}
+				} else {
+					podList := &v1.PodList{}
+					if err = p.azr.getClient().List(ctx, podList, &client.ListOptions{LabelSelector: podAntiAffinities}); err != nil {
+						klog.Errorf("failed to retrieve pod list: %v", err)
+						continue
+					}
+					pods = podList.Items
+				}
+
+				nodesWithSelectedPods := set{}
+				for _, pod := range pods {
+					nodesWithSelectedPods.add(pod.Spec.NodeName)
+				}
+
+				qualifiedLabelSet := map[string]set{}
+				for node := range nodesWithSelectedPods {
+					var nodeInstance v1.Node
+					if err = p.azr.getClient().Get(ctx, types.NamespacedName{Name: node.(string)}, &nodeInstance); err != nil {
+						klog.Errorf("failed to get node (%s)", node.(string))
+						continue
+					}
+					if value, exists := nodeInstance.Labels[podAntiAffinity.TopologyKey]; exists {
+						labelSet, exists := qualifiedLabelSet[value]
+						if !exists {
+							labelSet = set{}
+						}
+						labelSet.add(value)
+					}
+				}
+
+				selector := labels.NewSelector()
+				for key, values := range qualifiedLabelSet {
+					labelValues := values.toSlice().([]string)
+					requirements, err := azureutils.CreateLabelRequirements(key, selection.In, labelValues...)
+					if err != nil {
+						klog.Errorf("failed to create label for key (%s) and values (%+v) :%v", key, labelValues, err)
+						continue
+					}
+					selector = selector.Add(*requirements)
+				}
+
+				// remove any candidate node which does not satisfy the pod affinity
+				for candidateNode := range candidateNodes {
+					node := nodes[candidateNode.(int)]
+					nodeLabels := labels.Set(node.Labels)
+					if !selector.Matches(nodeLabels) {
+						klog.Infof("Removing node (%s) from candidate nodes: node does not satisfy inter-pod-affinity (%v), label (%v)", candidateNode, podAntiAffinity, selector)
+						candidateNodes.remove(candidateNode)
+					}
+				}
 			}
 		}
 	}
 
-	var volumeNodeSelectors []*nodeaffinity.NodeSelector
-	compatibleZonesSet := set{}
-	for i, volume := range volumes {
-		azVolume, err := azureutils.GetAzVolume(ctx, azr.getClient(), azr.getAzClient(), volume, azr.getSharedState().objectNamespace, true)
-		if err != nil {
-			klog.V(5).Infof("AzVolume for volume %s is not found.", volume)
-			return nil, err
+	filteredNodes := []v1.Node{}
+	for candidateNode := range candidateNodes {
+		if node, exists := nodeMap[candidateNode.(string)]; exists {
```

Comment:
Can't we just return the cadidateNodes without checking whether they are in the nodeMap or not. On that note, is there a need of having nodes as part of the filter struct? It seems sufficient to just have the candidate nodes in the filter function params.

Reply:
`nodeMap` is there so that we can map `node.Name` to object instance and this only applies to `inter-pod affinity filter`, where we need to add filtered nodes back into the candidateNode list based on where qualified pods' replica attachments are attached to, but I agree with you that we can get rid of node list from the `filter plugin` struct (this is remnant field from old implementation of the `filter plugin` that I deprecated, which I forgot to remove). Thanks for catching it 👍